lib/spack/spack/ci.py

```diff
@@ -546,7 +546,8 @@ def get_spec_filter_list(env, affected_pkgs, dependencies=True, dependents=True)
             if dependencies:
                 affected_specs.update(match.traverse(direction="children", root=False))
             if dependents:
-                affected_specs.update(match.traverse(direction="parents", root=False))
+                for dependent_spec in match.traverse(direction="parents", root=False):
+                    affected_specs.update(dependent_spec.traverse(direction="children", root=True))
```

Comment:
why traverse down in every intermediate parent?

Reply:
Because if a dependent of a touched specs turns out to be missing from the mirror (see the description), we'll schedule it for rebuilding.  If that dependent has any dependencies missing from the mirror too, we should schedule those as well, or we'll end up trying to build that dependent and we'll find the dependencies couldn't be installed from binary mirror when `--cache-only` was specified.

Reply:
If you do the latter, you can use the new api `traverse_nodes([root_a, root_b, ...])`

Reply:
> either use a visited set to avoid traversing the same edge you came from going up

Even though `affected_specs` is a `set`?

Reply:
Think of a dag that looks like `v1 <- v2 <- v3 <- ... <- vn` where `<-` points to the dependency. If you start with `v1` your traversal code will first iterate over `v1`, then `v2, v1`, then ... `v3, v2, v1`, ... `vn, ..., v1` which is quadratic and unnecessary. If you pass `traverse(visited=my_set)` it won't follow edges to nodes seen before.

Reply:
Also, I would just drop the keyword args `dependencies=True` and `dependents=True`, they're never set to anything else so there's no point to make it configurable, especially now that with this change `dependents` doesn't mean `dependents` and in fact implies `dependencies`.

Reply:
Oh, yes, now I get what you mean by "every intermediate parent".  Thanks for the suggestion.

Reply:
You may need to be a bit careful, cause the visited kwarg to `traverse` takes elements of whatever the `key` function produces (`id` by default) to distinguish unique specs. It's not a set of Spec instances.

Reply:
But I should be fine if provide `key=lambda s: s.dag_hash()` to `traverse`, as you suggested last week on slack?

Reply:
It will still be a set of strings (dag hashes), not specs:

```python
In [1]: cmake = spack.spec.Spec('cmake').concretized() # cmake

In [2]: child = cmake.dependencies()[0] # ncurses

In [3]: [s.name for s in cmake.traverse(visited=visited, key=lambda s: s.dag_hash())]
Out[3]: ['ncurses', 'pkgconf']

In [4]: visited
Out[4]: {'e5ewulkrwwvdck3b3dxsepgiujr7iqbz', 'xzo47hkvc77pabwn4kyyiwkjo3yud7xd'}

In [5]: [s.name for s in cmake.traverse(visited=visited, key=lambda s: s.dag_hash())]
Out[5]: ['cmake', 'openssl', 'ca-certificates-mozilla', 'perl', 'berkeley-db', 'bzip2', 'diffutils', 'libiconv', 'gdbm', 'readline', 'zlib']

In [6]: visited
Out[6]: {'5fie3rqgdtbik6qopaa5plhgfinuuywz', 'atlyybqxtsfjnci7xosrv4cczm7pjeh4', 'e5ewulkrwwvdck3b3dxsepgiujr7iqbz', 'fdvtzu2nvuktthv2eygftlywmxxmlq5p', 'gtwazdob6rf6mlcojjoartzucqx4r35x', 'hlnjdqnhzbz4sdg4b2nx3goh44iwb2qs', 'i55tsxsepukei4klbtquplznt5oxzyvy', 'm6gfpiqxuhtlryro5qt36sbp2xie5o7s', 'q27oxeoxa22tkhck7pbv2f6azjydzkia', 'vc4kuphzwsemw23up4o2kwx5tja5gvmq', 'w2gmdr7nimmcm6l25gbf3fe7mhqdathb', 'xzo47hkvc77pabwn4kyyiwkjo3yud7xd', 'y4zrxrek36joqs5hrg3hgwoko3esi5mw'}
```

but it's better to key by `dag_hash` in case you somehow end up with two `Spec` instances describing the exact same spec with the same hash (due to a copy or loading them from disk twice or so)

Reply:
I get it, thank you for providing the example.  I was confused at first because `In [4]` looks exactly the same as `In  [6]`, but I think it must have read `child.traverse(...)` when you evaluated it originally.

Reply:
Yeah, i changed it a few times 😆